### PR TITLE
Review/update documentation on environment modules

### LINKED ISSUE #155: Typo in worker framework section 10.1
@gjbex  The example `weather.pbs`   script includes these lines:

```...
# # This script is submitted to the cluster with the following 2 commands:
# module load worker
# wsub -data data.csv -batch weather.sh
...
```

but it should be changed by

```...
# # This script is submitted to the cluster with the following 2 commands:
# module load worker
# wsub -data data.csv -batch weather.pbs
...
```

### PR DESCRIPTION
**Pull request by redfast00**
Fixes  #155

### CODE REVIEW

File: intro-HPC/ch_multi_job_submission.tex [hunk 1]
```diff
@@ -130,7 +130,7 @@ \section{The worker Framework: Parameter Sweeps}
 The job can now be submitted as follows:
 
 \begin{prompt}
-%\shellcmd{module load worker}%
+%\shellcmd{module load worker/1.6.8-intel-2018a}%
```
Comment (boegel) [inline intro-HPC/ch_multi_job_submission.tex]: @redfast00 Is it mentioned somewhere how to check which `worker` modules are available (since what you mention here is site-specific)?

File: intro-HPC/ch_preparing_the_environment.tex [hunk 2]
```diff
@@ -1194,6 +1195,18 @@ \subsection{Explicit version numbers}
...
+example, to list the available versions of the GSL library:
+
+\begin{prompt}
+%\shellcmd{module avail GSL}%
+FGSL/1.1.0-intel-2016b (D)    GSL/2.3-intel-2016b
+GSL/2.1-intel-2016b           GSL/2.3-intel-2017a
+GSL/2.2.1-intel-2016b         GSL/2.4-GCCcore-6.4.0 (D)
+GSL/2.3-foss-2016b
+
+\end{prompt}
```
Comment (boegel) [inline intro-HPC/ch_preparing_the_environment.tex]: Why is this simply not a reference to the section about `module avail` we have already?

Also `GSL` is not the best example, since `FGSL` also pops up. At the very least explain this?

I'd also use `module avail GSL/`, to avoid partial match on anything module that includes `GSL` (that's particularly painful with `module avail R` if module is configured to be case-insensitive (which is site-specific))...

File: intro-HPC/ch_multi_job_submission.tex [hunk 3]
```diff
@@ -127,10 +127,11 @@ \section{The worker Framework: Parameter Sweeps}
 so the 100 calculations will be done in 1500/8 = 187.5 minutes, i.e., 4 hours
 to be on the safe side.
 
-The job can now be submitted as follows:
+The job can now be submitted as follows (to check what \verb|worker| version to use,
```
Comment (boegel) [inline intro-HPC/ch_multi_job_submission.tex]: `what \verb|worker| version` -> `which \verb|worker| module`

File: intro-HPC/ch_preparing_the_environment.tex (hunk 2 above)
Comment (redfast00) [reply intro-HPC/ch_preparing_the_environment.tex]: Because our current section about `module avail` doesn't talk about listing all versions: it only explains the `module avail` command that shows all modules.  Our Linux manual has a chapter (of one page) about the `module` command. Maybe that chapter should be moved to the HPC manual? (here's a link to it http://hpcugent.github.io/vsc_user_docs/pdf/intro-Linux-linux-gent.pdf#chapter.6)

Comment (boegel) [reply intro-HPC/ch_preparing_the_environment.tex]: I think it makes more sense to explain the `module` command in the HPC tutorial, since it's quite specific to HPC. See also https://github.com/hpcugent/vsc_user_docs/issues/192

**Comment by redfast00** [comment]
TODO: add explanation about explicit version numbers

**Comment by redfast00** [comment]
Note to self: use imaginary software names and versions

**Comment by boegel** [comment]
@redfast00 Hmm, looks like this needs a sync with `master` (diff looks incorrect on GitHub)?

File: intro-HPC/ch_preparing_the_environment.tex [hunk 4]
```diff
@@ -822,560 +822,3 @@ \section{Transfer Files to/from the \hpc}
 \subsection{Fast file transfer for large datasets}
 
 See \href{\LinuxManualURL#sec:rsync}{the section on \texttt{rsync} in chapter 5 of the Linux intro manual}.
-
-\section{Modules}
-\label{sec:modules}
```
Comment (boegel) [inline intro-HPC/ch_preparing_the_environment.tex]: Any references that got broken?

Can we easily detect broken references with a `grep` in Travis (if so, should be done in a separate PR, not here)

File: intro-HPC/ch_running_batch_jobs.tex [hunk 5]
```diff
@@ -40,6 +40,331 @@ \chapter{Running batch jobs}
...
+\item  Configure look and feel of the shell.
+\item  Setup terminal settings depending on which terminal you're using.
+\item  Set the search path for running executables.
+\item  Set environment variables as needed by programs.
+\item  Set convenient abbreviations for heavily used values.
+\item  Run commands that you want to run whenever you log in or log out.
+\item  Setup aliases and/or shell function to automate tasks to save typing and time.
+\item  Changing the bash prompt.
+\item  Setting shell options.
+\end{enumerate}
```
Comment (boegel) [inline intro-HPC/ch_running_batch_jobs.tex]: @redfast00 This is probably way too much detail, especially for people that already know what environment variables are, it looks out of place in a section on modules.

Trim this down significantly, shouldn't be more than a couple of sentences what environment variables are (not what a shell is...) + a pointer to the Linux intro for more information?

File: intro-HPC/ch_running_batch_jobs.tex [hunk 6]
```diff
@@ -40,6 +40,331 @@ \chapter{Running batch jobs}
...
+The environment variables are typically set at login by a script, whenever you
+connect to the \hpc. These pre-defined variables usually impact the run time
+behaviour of the programs that we want to run.
+
+All the software packages that are installed on the \hpc cluster require
+different settings. These packages include compilers, interpreters,
+mathematical software such as MATLAB and SAS, as well as other applications and
+libraries.
+
+In order to administer the active software and their environment variables, a
```
Comment (boegel) [inline intro-HPC/ch_running_batch_jobs.tex]: new `subsection`: "The \verb|module| command"

File: intro-HPC/ch_running_batch_jobs.tex [hunk 7]
```diff
@@ -40,6 +40,331 @@ \chapter{Running batch jobs}
...
+\strong{\emph{``module''}} package has been developed, which:
```
Comment (boegel) [inline intro-HPC/ch_running_batch_jobs.tex]: the tool/concept is actually called "environment modules" (cfr. http://modules.sourceforge.net/, but don't include that link since some sites using a different "fork", Lmod, see https://lmod.readthedocs.io), use that term once and then mention we'll refer to it as "module system" for short

File: intro-HPC/ch_preparing_the_environment.tex (hunk 4 above)
Comment (redfast00) [reply intro-HPC/ch_preparing_the_environment.tex]: References didn't get broken, since I didn't remove any labels (just moved them).

File: intro-HPC/ch_running_batch_jobs.tex [hunk 8]
```diff
@@ -40,6 +40,331 @@ \chapter{Running batch jobs}
...
+
+\begin{enumerate}
+\item  Activates or deactivates \underbar{software packages} and their dependencies.
```
Comment (boegel) [inline intro-HPC/ch_running_batch_jobs.tex]: Replace `\underbar` with `\emph`?

File: intro-HPC/ch_running_batch_jobs.tex [hunk 9]
```diff
@@ -40,6 +40,331 @@ \chapter{Running batch jobs}
...
+\item  Allows setting and unsetting of \underbar{environment variables}, including adding and deleting entries from database-type environment variables.
+\item  Does this in a \underbar{shell-independent} fashion (necessary information is stored in the accompanying module configuration file).
```
Comment (boegel) [inline intro-HPC/ch_running_batch_jobs.tex]: `module configuration file` -> `module file`

File: intro-HPC/ch_running_batch_jobs.tex [hunk 10]
```diff
@@ -40,6 +40,331 @@ \chapter{Running batch jobs}
...
+\item  Takes care of \underbar{versioning aspects}: For many libraries, multiple versions are installed and maintained. The module system also takes care of the versioning of software packages in case multiple versions are installed. For instance, it does not allow multiple versions to be loaded at same time.
```
Comment (boegel) [inline intro-HPC/ch_running_batch_jobs.tex]: partial overlap between 1st & 2nd sentence? reads a bit weird... Just drop the `in case multiple versions are installed` in the 2nd sentence?

File: intro-HPC/ch_running_batch_jobs.tex [hunk 11]
```diff
@@ -40,6 +40,331 @@ \chapter{Running batch jobs}
...
+\item  Takes care of \underbar{dependencies}: Another issue arises when one considers library versions and the dependencies they create. Some software requires an older version of a particular library to run correctly (or at all). Hence a variety of version numbers is available for important libraries.
```
Comment (boegel) [inline intro-HPC/ch_running_batch_jobs.tex]: `create` -> `require`

Comment (boegel) [reply intro-HPC/ch_running_batch_jobs.tex]: Also mention that modules typically load (modules for) the required dependencies automatically.

File: intro-HPC/ch_running_batch_jobs.tex [hunk 12]
```diff
@@ -40,6 +40,331 @@ \chapter{Running batch jobs}
...
+
+\begin{prompt}
+%\shellcmd{module av}%
+\end{prompt}
+
+or by typing it in full:
+
+\begin{prompt}
+%\shellcmd{module available}%
+\end{prompt}
```
Comment (boegel) [inline intro-HPC/ch_running_batch_jobs.tex]: There are multiple equivalent commands: `module available` (that's the long version), `module avail`, `module av` also work.

With Lmod (UGent-specific), also `ml av` & the longer ones using `ml` rather than `module` work.

Please try to clarify that, but keep it short...

File: intro-HPC/ch_running_batch_jobs.tex [hunk 13]
```diff
@@ -40,6 +40,331 @@ \chapter{Running batch jobs}
...
+
+This will give some output such as:
+
+\inputsite{available-modules}
+
+This gives a full list of software packages that can be loaded. Note that
+modules starting with a capital letter are listed first.
```
Comment (boegel) [inline intro-HPC/ch_running_batch_jobs.tex]: drop the last sentence, Lmod doesn't do that

You can mention that the casing of module names is important (so lowercase vs uppercase letter matter in module names).

File: intro-HPC/ch_running_batch_jobs.tex [hunk 14]
```diff
@@ -40,6 +40,331 @@ \chapter{Running batch jobs}
...
+
+\subsection{Organisation of modules in toolchains}
+
+The amount of modules on the VSC systems can be overwhelming, and it is not
+always immediately clear which modules can be loaded safely together if you
+need to combine multiple programs in a single job to get your work done.
+
+Therefore the VSC has defined so-called \strong{toolchains} on the newer VSC-clusters.
```
Comment (boegel) [inline intro-HPC/ch_running_batch_jobs.tex]: drop `on the newer VSC-clusters.`

File: intro-HPC/ch_running_batch_jobs.tex [hunk 15]
```diff
@@ -40,6 +40,331 @@ \chapter{Running batch jobs}
...
+A toolchain contains a C/C++ and Fortran compiler, a MPI library and some basic
+math libraries for (dense matrix) linear algebra and FFT. Two toolchains are
+defined on most VSC systems. One, the \lstinline|intel| toolchain, consists of the Intel
+compilers, MPI library and math libraries. The other one, the \lstinline|foss| toolchain,
+consists of Open Source components: the GNU compilers, OpenMPI, OpenBLAS and
+the standard LAPACK and ScaLAPACK libraries for the linear algebra operations
+and the FFTW library for FFT. The toolchains are refreshed twice a year,
+which is reflected in their name.
+
+E.g., \lstinline|foss-2014a| is the first version of the \lstinline|foss| toolchain in 2014.
```
Comment (boegel) [inline intro-HPC/ch_running_batch_jobs.tex]: update to `2018` :), and maybe use `\year` here if something like that exists in LaTeX

File: intro-HPC/ch_running_batch_jobs.tex [hunk 16]
```diff
@@ -40,6 +40,331 @@ \chapter{Running batch jobs}
...
+together well without conflicts.
+
+For some clusters, additional toolchains are defined, e.g., to take advantage
+of specific properties of that cluster such as GPU accelerators or special
+interconnect features that require a vendor-specific MPI-implementation.
+
+\subsection{Activating and de-activating modules}
+\label{subsec:activating-and-deactivating-modules}
+
+A module is loaded using the following command:
```
Comment (boegel) [inline intro-HPC/ch_running_batch_jobs.tex]: first use `activate`, then switch to `load` and keep using `load` afterwards

"To 'activate' a software package, you load the corresponding module file using the `module load` command. (or with Lmod, `ml` (which is short for both `module load` and `module list`, genius...)

File: intro-HPC/ch_running_batch_jobs.tex [hunk 17]
```diff
@@ -40,6 +40,331 @@ \chapter{Running batch jobs}
...
+
+
+\begin{prompt}
+%\shellcmd{module load \emph{examplemodule}}%
+\end{prompt}
+
+This will load the most recent version of \emph{examplemodule}.
```
Comment (boegel) [inline intro-HPC/ch_running_batch_jobs.tex]: Clarify that this is bad practice usually: you should  almost always load a *specific* version of a module, rather than just loading the default version (which usually corresponds to the latest version). The default version can change over time, and so you'll suddenly be loading a different software version, which can break your job script or result in different results...

File: intro-HPC/ch_running_batch_jobs.tex [hunk 18]
```diff
@@ -40,6 +40,331 @@ \chapter{Running batch jobs}
...
+
+For some packages (e.g., OpenMPI), multiple versions are installed; the load
```
Comment (boegel) [inline intro-HPC/ch_running_batch_jobs.tex]: drop `OpenMPI`, we have multiple versions of pretty much everything

File: intro-HPC/ch_running_batch_jobs.tex [hunk 19]
```diff
@@ -40,6 +40,331 @@ \chapter{Running batch jobs}
...
+command will automatically choose the most recent version (i.e., the
+lexicographical last after the ``/'') or the default version (as set by the
+system administrators). However, the user can (and probably should, to avoid
```
Comment (boegel) [inline intro-HPC/ch_running_batch_jobs.tex]: It's the other way around: default version if it was set, most recent version otherwise.

File: intro-HPC/ch_running_batch_jobs.tex (hunk 17 above)
Comment (boegel) [reply intro-HPC/ch_running_batch_jobs.tex]: Ah, that's actually done below... Should maybe stand out more?

File: intro-HPC/ch_running_batch_jobs.tex [hunk 20]
```diff
@@ -40,6 +40,331 @@ \chapter{Running batch jobs}
...
+
+Obviously, you need to keep track of the modules that are currently
+loaded. If you executed the two load commands stated above, you will get the
+following:
+
+\begin{prompt}
+%\shellcmd{module list}%
+Currently Loaded Modulefiles:
+  1) examplemodule/1.2.3-intel-2018a        3) intel/2018a
+  2) secondexample/2.7-intel-2018a          4) examplelib/1.2-intel-2014b
```
Comment (boegel) [inline intro-HPC/ch_running_batch_jobs.tex]: This can never happen, you can never load modules with two different `intel` versions.

Also, in reality you'll have over a dozen modules loaded as soon as you load something that also loads the `intel` toolchain...

Show that, and explain why you suddenly have a whole bunch of modules loaded.

This is a reason *not* to use `\year`, since the versions of the modules loaded by `intel` depend heavily on the `intel` version...

File: intro-HPC/ch_running_batch_jobs.tex [hunk 21]
```diff
@@ -40,6 +40,331 @@ \chapter{Running batch jobs}
...
+\end{prompt}
+
+
+It is important to note at this point that other modules (e.g., \lstinline|intel/2014a|)
```
Comment (boegel) [inline intro-HPC/ch_running_batch_jobs.tex]: `2018`, use `\year` to keep it both updated & consistent

File: intro-HPC/ch_running_batch_jobs.tex [hunk 22]
```diff
@@ -40,6 +40,331 @@ \chapter{Running batch jobs}
...
+\begin{prompt}
+%\shellcmd{module load \emph{secondexample/2.7-intel-2018a}}%
+\end{prompt}
+
+Obviously, you need to keep track of the modules that are currently
+loaded. If you executed the two load commands stated above, you will get the
+following:
+
+\begin{prompt}
+%\shellcmd{module list}%
```
Comment (boegel) [inline intro-HPC/ch_running_batch_jobs.tex]: or, equivalent `ml` with Lmod

File: intro-HPC/ch_running_batch_jobs.tex [hunk 23]
```diff
@@ -40,6 +40,331 @@ \chapter{Running batch jobs}
...
+and the system administrator specified that the \lstinline|intel/2014a| module should
+be loaded whenever the \lstinline|secondexample| module is loaded. There are advantages and
+disadvantages to this, so be aware of automatically loaded modules whenever
+things go wrong: they may have something to do with it!
+
+In fact, an easy way to check the components and version numbers of those
+components of a toolchain is to simply load the toolchain and then list the
+modules that are loaded.
+
+To unload a module, one can use the \lstinline|module unload| command. It works
```
Comment (boegel) [inline intro-HPC/ch_running_batch_jobs.tex]: or `ml -example` with Lmod , equivalent with `module unload example`

File: intro-HPC/ch_running_batch_jobs.tex [hunk 24]
```diff
@@ -40,6 +40,331 @@ \chapter{Running batch jobs}
...
+consistently with the load command, and reverses the latter's effect. However,
+the dependencies of the package are NOT automatically unloaded; the user shall
```
Comment (boegel) [inline intro-HPC/ch_running_batch_jobs.tex]: `the user shall unload` -> `you will have to unload`

File: intro-HPC/ch_running_batch_jobs.tex [hunk 25]
```diff
@@ -40,6 +40,331 @@ \chapter{Running batch jobs}
...
+unload the packages one by one. One can however unload automatically loaded
+modules manually, to debug some problem. When the \lstinline|secondexample| module is unloaded,
```
Comment (boegel) [inline intro-HPC/ch_running_batch_jobs.tex]: drop the `One can however unload automatically loaded modules manually, to debug some problem.`

File: intro-HPC/ch_running_batch_jobs.tex [hunk 26]
```diff
@@ -40,6 +40,331 @@ \chapter{Running batch jobs}
...
+\begin{prompt}
+%\shellcmd{module unload \emph{secondexample}}%
+%\shellcmd{module list}%
+Currently Loaded Modulefiles:
+  1) examplemodule/1.2.3-intel-2018a        3) intel/2018a
+  2) examplelib/1.2-intel-2014b
+\end{prompt}
+
+Notice that the version was not specified: the module system is sufficiently
+clever to figure out what the user intends. However, checking the list of
```
Comment (boegel) [inline intro-HPC/ch_running_batch_jobs.tex]: it's not clever, you can only have one version loaded of each module, so there's no ambiguity when only specifying the module name when unloading...

File: intro-HPC/ch_running_batch_jobs.tex [hunk 27]
```diff
@@ -40,6 +40,331 @@ \chapter{Running batch jobs}
...
+currently loaded modules is always a good idea, just to make sure \ldots
```
Comment (boegel) [inline intro-HPC/ch_running_batch_jobs.tex]: drop the `just to make sure...`

File: intro-HPC/ch_running_batch_jobs.tex [hunk 28]
```diff
@@ -40,6 +40,331 @@ \chapter{Running batch jobs}
...
+
+\begin{prompt}
+%\shellcmd{module purge}%
+\end{prompt}
+
+However, on some VSC clusters you may be left with a very empty list of available
+modules after executing \lstinline|module purge|. On those systems, \lstinline|module av| will show you a list of
+modules containing the name of a cluster or a particular feature of a section of
+the cluster, and loading the appropriate module will restore the module list
+applicable to that particular system.
```
Comment (boegel) [inline intro-HPC/ch_running_batch_jobs.tex]: This needs to be rewritten. In Ghent, `module purge` is always safe, the `cluster` module will not get unloaded (because it's a "sticky" module), only `module --force purge` will get you in trouble (no need to mention that).

File: intro-HPC/ch_running_batch_jobs.tex [hunk 29]
```diff
@@ -40,6 +40,331 @@ \chapter{Running batch jobs}
...
+
+Modules need not be loaded one by one; the two \lstinline|module load| commands can be
+combined as follows:
+
+\begin{prompt}
+%\shellcmd{module load \emph{examplemodule/1.2.3-intel-2018a} \emph{secondexample/2.7-intel-2018a}}%
+\end{prompt}
+
+This will load the two modules as well as their dependencies.
```
Comment (boegel) [inline intro-HPC/ch_running_batch_jobs.tex]: this last bit belongs in the `module load` section

File: intro-HPC/ch_running_batch_jobs.tex [hunk 30]
```diff
@@ -40,6 +40,331 @@ \chapter{Running batch jobs}
...
+rather than
+
+\begin{prompt}
+%\shellcmd{module load \emph{examplemodule/1.2.3-intel-2018a}}%
+\end{prompt}
+
+Everything works fine, up to the point where a new version of \lstinline|examplemodule| is installed,
+4.5.6 compiled for gcc. From then on, the user's load command will load the
+latter version, rather than the intended one, which may lead to unexpected
+problems.
```
Comment (boegel) [inline intro-HPC/ch_running_batch_jobs.tex]: refer to troubleshooting section here?

File: intro-HPC/ch_running_batch_jobs.tex [hunk 31]
```diff
@@ -40,6 +40,331 @@ \chapter{Running batch jobs}
...
+using:
+
+\begin{prompt}
+%\shellcmd{module load \emph{examplemodule}}%
+\end{prompt}
+
+rather than
+
+\begin{prompt}
+%\shellcmd{module load \emph{examplemodule/1.2.3-intel-2018a}}%
```
Comment (boegel) [inline intro-HPC/ch_running_batch_jobs.tex]: I would use a single `example` module here with 2 versions `1.2.3` and `4.5.6` (no `intel` part in it, that's irrelevant here)

File: intro-HPC/ch_running_batch_jobs.tex [hunk 32]
```diff
@@ -40,6 +40,331 @@ \chapter{Running batch jobs}
...
+examplemodule/1.2.3-foss-2018a
+
+\end{prompt}
+
+Lets now generate a version conflict with the \lstinline|thirdexample| module, and see what
+is happening.
+
+\ifusinglmod
+\begin{prompt}
+%\shellcmd{module av thirdexample/}%
```
Comment (boegel) [inline intro-HPC/ch_running_batch_jobs.tex]: just use `example` here, no need to switch to something else?

File: intro-HPC/ch_running_batch_jobs.tex [hunk 33]
```diff
@@ -40,6 +40,331 @@ \chapter{Running batch jobs}
...
+To get a list of all possible commands, type:
+
+\begin{prompt}
+%\shellcmd{module help}%
+\end{prompt}
+
+Or to get more information about one specific module package:
+
+\begin{prompt}
+%\shellcmd{module help examplemodule}%
```
Comment (boegel) [inline intro-HPC/ch_running_batch_jobs.tex]: use specific version, `example/1.2.3`

File: intro-HPC/ch_running_batch_jobs.tex [hunk 34]
```diff
@@ -40,6 +40,331 @@ \chapter{Running batch jobs}
...
+----------- Module Specific Help for 'examplemodule/1.2.3-intel-2018a' ---------------------------
+
+   This is a module that doesn't really exist. It can be used to do things. - Homepage: https://example.com/
```
Comment (boegel) [inline intro-HPC/ch_running_batch_jobs.tex]: This is just an example - Homepage: https://example.com

File: intro-HPC/ch_running_batch_jobs.tex [hunk 35]
```diff
@@ -40,6 +40,331 @@ \chapter{Running batch jobs}
...
+\ifusinglmod
+\subsection{Save and load collections of modules}
+If you have a set of modules that you need to load often,
+you can save these in a \emph{collection}. This will enable you to load all the
+module you need with a single command.
+
+First, load all modules you want to include in the collections:
+
+\begin{prompt}
+%\shellcmd{module load \emph{examplemodule/1.2.3-intel-2018a} \emph{secondexample/2.7-intel-2018a}}%
```
Comment (boegel) [inline intro-HPC/ch_running_batch_jobs.tex]: again use modules here without `intel` in the version, keep it nice & short

File: intro-HPC/ch_running_batch_jobs.tex [hunk 36]
```diff
@@ -40,6 +40,331 @@ \chapter{Running batch jobs}
...
+\end{prompt}
+
+Now store it in a collection using \lstinline|ml save|. In this example, the collection
```
Comment (boegel) [inline intro-HPC/ch_running_batch_jobs.tex]: `module save` or `ml save`

File: intro-HPC/ch_running_batch_jobs.tex [hunk 37]
```diff
@@ -40,6 +40,331 @@ \chapter{Running batch jobs}
...
+is named \lstinline|my-collection|.
+
+\begin{prompt}
+%\shellcmd{module save \emph{my-collection}}%
+\end{prompt}
+
+Later, for example in a jobscript or a new session, you can load all these modules with
+\lstinline|module restore|:
```
Comment (boegel) [inline intro-HPC/ch_running_batch_jobs.tex]: or `ml restore` (do test)

File: intro-HPC/ch_preparing_the_environment.tex (hunk 4 above)
Comment (boegel) [reply intro-HPC/ch_preparing_the_environment.tex]: OK, but please open an issue to add a check in Travis to check for broken references, I don't think that makes the LaTeX build fail, does it?

**Comment by redfast00** [comment]
TODO: 

- [ ] rewrite `module purge`
- [ ] ml spider
- [ ] ml show

File: intro-HPC/ch_running_batch_jobs.tex [hunk 38]
```diff
@@ -40,6 +40,338 @@ \chapter{Running batch jobs}
...
+loaded. If you executed the two load commands stated above, you will get the
+following:
+
+\begin{prompt}
+%\shellcmd{module list}%
+Currently Loaded Modulefiles:
+  1) example/1.2.3-intel-2018a        5) intel/2018a
+  2) secondexample/2.7-intel-2018a          6) examplelib/1.2-intel-2018b
+  3) GCCcore/4.9.3                          7) binutils/2.25-GCCcore-4.9.3
+  4) ifort/2018.1.150-GCC-4.9.3-2.25        8) %\ldots{}%
```
[the remark this answers is not in the document]
Comment (redfast00) [reply intro-HPC/ch_running_batch_jobs.tex]: I just discovered that the order the modules are listed in is the order they have been loaded :)

File: intro-HPC/ch_running_batch_jobs.tex [hunk 39]
```diff
@@ -40,6 +40,347 @@ \chapter{Running batch jobs}
...
+In order to unload all modules at once, and hence be sure to start in a clean
+state, you can use:
+
+\begin{prompt}
+%\shellcmd{module purge}%
+\end{prompt}
+
+\ifgent
+This is always safe: the \lstinline|cluster| module (the module that specifies what cluster
+jobs will get submitted to) will not be unloaded (because it's a ``sticky'' module).
```
[the remark this answers is not in the document]
Comment (wpoely86) [reply intro-HPC/ch_running_batch_jobs.tex]: We don't have cluster modules, so no.

File: intro-HPC/ch_preparing_the_environment.tex (hunk 4 above)
Comment (redfast00) [reply intro-HPC/ch_preparing_the_environment.tex]: Done, see #295

File: intro-HPC/ch_running_batch_jobs.tex [hunk 40]
```diff
@@ -40,6 +40,494 @@ \chapter{Running batch jobs}
...
+    variety of version numbers is available for important libraries. Modules
+    typically load the required dependencies automatically.
+\end{enumerate}
+
+This is all managed with the \lstinline|module| command, which is explained in the next sections.
+
+\ifusinglmod
+There is also a shorter \lstinline|ml| command that does exactly the same as the \lstinline|module|
+command and is easier to type. Whenever you see a \lstinline|module| command, you can replace
+\lstinline|module| with \lstinline|ml|.s
```
Comment (boegel) [inline intro-HPC/ch_running_batch_jobs.tex]: what's with the trailing `s`? slip of the finger?

File: intro-HPC/ch_running_batch_jobs.tex [hunk 41]
```diff
@@ -40,6 +40,494 @@ \chapter{Running batch jobs}
...
+and the FFTW library for FFT. The toolchains are refreshed twice a year,
+which is reflected in their name.
+
+E.g., \texttt{foss/\the\year{}a} is the first version of the \lstinline|foss| toolchain in \the\year.
+
+The toolchains are then used to compile a lot of the software installed on
+the VSC clusters. You can recognise those packages easily as they all contain the
+name of the toolchain after the version number in their name (e.g. \lstinline|Python/2.7.12-intel-2016b|).
+Packages compiled with the same toolchain and toolchain version will typically work
+together well without conflicts.
```
Comment (boegel) [inline intro-HPC/ch_running_batch_jobs.tex]: That's actually not true at all, since there often are conflicts on the dependencies as well, e.g. Python 2 vs 3. We pay more attention to this recently in Ghent, but only for the most recent toolchains (`2018a` & newer)

Please rephrase: `Only packages compiled with the same toolchain name and version can work together without conflicts.`

File: intro-HPC/ch_running_batch_jobs.tex [hunk 42]
```diff
@@ -40,6 +40,494 @@ \chapter{Running batch jobs}
...
+\end{prompt}
+
+Everything works fine, up to the point where a new version of \lstinline|example| is installed, 4.5.6.
+From then on, the user's \lstinline|load| command will load the latter version, rather than
+the intended one, which may lead to unexpected problems.
+\ifgent % The troubleshooting section is wrapped in an ifgent, we don't want undefined references when compiling for antwerpen,...
+See for example \autoref{sec:module-conflicts}.
+\fi
+
+Consider the followin \lstinline|example| modules:
```
Comment (boegel) [inline intro-HPC/ch_running_batch_jobs.tex]: `following` :)

File: intro-HPC/ch_running_batch_jobs.tex [hunk 43]
```diff
@@ -40,6 +40,494 @@ \chapter{Running batch jobs}
...
+\end{prompt}
+\fi
+
+To remove a collection, remove the corresponding file in \lstinline|$HOME/.lmod.d|:
+
+\begin{prompt}
+%\shellcmd{rm \$HOME/.lmod.d/my-collection}%
+\end{prompt}
+
+\ifusinglmod
```
Comment (boegel) [inline intro-HPC/ch_running_batch_jobs.tex]: `module show` is not specific to Lmod (`module spider` is though), so this shouldn't be guarded

File: intro-HPC/ch_running_batch_jobs.tex [hunk 44]
```diff
@@ -40,6 +40,494 @@ \chapter{Running batch jobs}
...
+prepend_path("ACLOCAL_PATH","/apps/gent/CO7/sandybridge/software/GSL/2.4-GCCcore-6.4.0/share/aclocal")
+prepend_path("CPATH","/apps/gent/CO7/sandybridge/software/GSL/2.4-GCCcore-6.4.0/include")
+prepend_path("LD_LIBRARY_PATH","/apps/gent/CO7/sandybridge/software/GSL/2.4-GCCcore-6.4.0/lib")
+prepend_path("LIBRARY_PATH","/apps/gent/CO7/sandybridge/software/GSL/2.4-GCCcore-6.4.0/lib")
+prepend_path("MANPATH","/apps/gent/CO7/sandybridge/software/GSL/2.4-GCCcore-6.4.0/share/man")
+prepend_path("PATH","/apps/gent/CO7/sandybridge/software/GSL/2.4-GCCcore-6.4.0/bin")
+prepend_path("PKG_CONFIG_PATH","/apps/gent/CO7/sandybridge/software/GSL/2.4-GCCcore-6.4.0/lib/pkgconfig")
+setenv("EBROOTGSL","/apps/gent/CO7/sandybridge/software/GSL/2.4-GCCcore-6.4.0")
+setenv("EBVERSIONGSL","2.4")
+setenv("EBDEVELGSL","/apps/gent/CO7/sandybridge/software/GSL/2.4-GCCcore-6.4.0/easybuild/GSL-2.4-GCCcore-6.4.0-easybuild-devel")
```
Comment (boegel) [inline intro-HPC/ch_running_batch_jobs.tex]: I would i) use the output of the `module show Python/...` so you can point out that `Python` is usually installed with additional extensions, ii) clean up the output a bit to only show what's relevant.

So, only include one single `load` statement (e.g. for the toolchain), replace the others with a single `...`.
Also only include a single `prepend_path` statement, like this: `prepend_path(...)`.

When showing the output, you still need to discriminate between Lmod & not.
Log in to Leuven to check output without Lmod (and also trim it down).

File: intro-HPC/ch_running_batch_jobs.tex [hunk 45]
```diff
@@ -40,6 +40,494 @@ \chapter{Running batch jobs}
...
+================
+ - Homepage: http://www.gnu.org/software/gsl/
+
+]])
+\end{prompt}
+\fi
+
+
+If you're not sure what all of this means: don't worry, you don't have to know;
+just load the module and try to use the software.
```
Comment (boegel) [inline intro-HPC/ch_running_batch_jobs.tex]: Just highlight how people can see which extensions/extra Python packages are installed as a part of Python...

**Comment by boegel** [comment]
looks good to go, but Travis fails?
